packages/autoloader/src/class-version-loader.php

```diff
@@ -0,0 +1,156 @@
+<?php
+/* HEADER */ // phpcs:ignore
+
+/**
+ * This class loads other classes based on given parameters.
+ */
+class Version_Loader {
+
+	/**
+	 * The Version_Selector object.
+	 *
+	 * @var Version_Selector
+	 */
+	private $version_selector;
+
+	/**
+	 * A map of available classes and their version and file path.
+	 *
+	 * @var array
+	 */
+	private $classmap;
+
+	/**
+	 * A map of PSR-4 namespaces and their version and directory path.
+	 *
+	 * @var array
+	 */
+	private $psr4_map;
+
+	/**
+	 * A map of all the files that we should load.
+	 *
+	 * @var array
+	 */
+	private $filemap;
+
+	/**
+	 * The constructor.
+	 *
+	 * @param Version_Selector $version_selector The Version_Selector object.
+	 * @param array            $classmap The verioned classmap to load using.
+	 * @param array            $psr4_map The versioned PSR-4 map to load using.
+	 * @param array            $filemap The versioned filemap to load.
+	 */
+	public function __construct( $version_selector, $classmap, $psr4_map, $filemap ) {
+		$this->version_selector = $version_selector;
+		$this->classmap         = $classmap;
+		$this->psr4_map         = $psr4_map;
+		$this->filemap          = $filemap;
+	}
+
+	/**
+	 * Finds the file path for the given class.
+	 *
+	 * @param string $class_name The class to find.
+	 *
+	 * @return string|null $file_path The path to the file if found, null if no class was found.
+	 */
+	public function find_class_file( $class_name ) {
+		$data = $this->find_newest_file(
+			isset( $this->classmap[ $class_name ] ) ? $this->classmap[ $class_name ] : null,
+			$this->find_psr4_file( $class_name )
```

Comment:
What do you think of checking whether `$this->psr4_map` is set here in `find_class_file`? If it's not set, we can return the file path from the classmap (if it exists), and we wouldn't need to call `find_newest_file()` or `find_psr4_file()`. Since the PSR-4 map won't be used in production, the package class file paths will be obtained from the classmap most of the time.

Reply:
We discussed this and decided not to make any changes because they would reduce code readability.